### ISSUE TITLE
Error when handing eth_call

### ISSUE THREAD

**Issue by MysticRyuujin** [NethermindEth/nethermind#2098]
I was doing some more testing, trying to get MyCrypto to work (still), and I came across this error on the node while scanning for tokens.

```
2020-07-08 22:22:46.4829|Error when handling ID a23133a1eda39e3490a57c1c20ee9089, eth_call({
  "to": "0x614b9802D45Aa1bC2282651dC1408632F9027A6e",
  "data": "0x70a08231000000000000000000000000338e2deb1e65238d1475e0c8b3d1b3cdfa5656cc"
},pending) | {"jsonrpc":"2.0","error":{"code":-32015,"message":"VM execution error.","data":"revert"},"id":"a23133a1eda39e3490a57c1c20ee9089"}
2020-07-08 22:22:47.1311|Error when handling ID 86d89d11608ade89e2bc32807652e99a, eth_call({
  "to": "0x8F936fE0faF0604c9C0Ef2406bde0A65365515d6",
  "data": "0x70a08231000000000000000000000000338e2deb1e65238d1475e0c8b3d1b3cdfa5656cc"
},pending) | {"jsonrpc":"2.0","error":{"code":-32015,"message":"VM execution error.","data":"revert"},"id":"86d89d11608ade89e2bc32807652e99a"}
2020-07-08 22:22:47.6743|Error when handling ID 0779d62a4c238b59baa7a46e3a0db07b, eth_call({
  "to": "0x1ed2B1eaEd8e968bc36EB90a914660A71827A5E9",
  "data": "0x70a08231000000000000000000000000338e2deb1e65238d1475e0c8b3d1b3cdfa5656cc"
},pending) | {"jsonrpc":"2.0","error":{"code":-32015,"message":"VM execution error.","data":"revert"},"id":"0779d62a4c238b59baa7a46e3a0db07b"}
2020-07-08 22:22:47.8519|Error when handling ID ea70a1f0a21aab8d9713db1383a936d2, eth_call({
  "to": "0x614b9802D45Aa1bC2282651dC1408632F9027A6e",
  "data": "0x70a08231000000000000000000000000338e2deb1e65238d1475e0c8b3d1b3cdfa5656cc"
},pending) | {"jsonrpc":"2.0","error":{"code":-32015,"message":"VM execution error.","data":"revert"},"id":"ea70a1f0a21aab8d9713db1383a936d2"}
2020-07-08 22:22:48.0346|Error when handling ID 73efb6a5c4ad036fde5dfe8309a419ee, eth_call({
  "to": "0x1ed2B1eaEd8e968bc36EB90a914660A71827A5E9",
  "data": "0x70a08231000000000000000000000000338e2deb1e65238d1475e0c8b3d1b3cdfa5656cc"
},pending) | {"jsonrpc":"2.0","error":{"code":-32015,"message":"VM execution error.","data":"revert"},"id":"73efb6a5c4ad036fde5dfe8309a419ee"}
2020-07-08 22:22:49.0937|Sync peers 60(61)/300
2020-07-08 22:22:53.0713|Error when handling ID d168ea46ded54672af6d42c83ae468ab, eth_call({
  "to": "0x614b9802D45Aa1bC2282651dC1408632F9027A6e",
  "data": "0x70a08231000000000000000000000000338e2deb1e65238d1475e0c8b3d1b3cdfa5656cc"
},pending) | {"jsonrpc":"2.0","error":{"code":-32015,"message":"VM execution error.","data":"revert"},"id":"d168ea46ded54672af6d42c83ae468ab"}
2020-07-08 22:22:53.1717|Error when handling ID dc4382eb6a310547225b4a41e717812b, eth_call({
  "to": "0x8F936fE0faF0604c9C0Ef2406bde0A65365515d6",
  "data": "0x70a08231000000000000000000000000338e2deb1e65238d1475e0c8b3d1b3cdfa5656cc"
},pending) | {"jsonrpc":"2.0","error":{"code":-32015,"message":"VM execution error.","data":"revert"},"id":"dc4382eb6a310547225b4a41e717812b"}
```

**Comment by MysticRyuujin** [comment]
Here are 2 JSON RPC Debug log files from my 2 Nethermind nodes (they're load balanced).

This log contains all calls made my MyCrypto from the moment I hit "Scan Tokens" until it errored out on the website.

[debug01.log](https://github.com/NethermindEth/nethermind/files/4896814/debug01.log)
[debug02.log](https://github.com/NethermindEth/nethermind/files/4896815/debug02.log)

**Comment by MysticRyuujin** [comment]
I've received more reports in the ArchiveNode.io Discord that suggests Nethermimd is not properly handling eth_call when there is a revert. So that leads me to suspect that is the root cause? Also it was pointed out that, in this log, the string "pending" is not wrapped in quotes, not sure if a logging issue or pointing to a possible cause?

**Comment by tkstanczak** [comment]
Logging is unrelated.

**Comment by tkstanczak** [comment]
We do not have 'pending' support and I think we just replace it with latest.

**Comment by tkstanczak** [comment]
sorry, closed by mistake

**Comment by tkstanczak** [comment]
We will need to compare reverts results as there may be difference between what Geth and what Parity do and we would have to have a switch to follow one pattern. Not sure yet. Needs some investigation.

**Comment by MysticRyuujin** [comment]
Whatever Geth and Parity do, I'm pretty sure they don't throw an error and timeout 😉 haha

**Comment by SebastianDremo** [comment]
Hey, just to let you know that I am working on this atm. 

Will inform you @MysticRyuujin if I found out what's causing the issue

**Comment by cburgdorf** [comment]
To give you an example of a hanging API call:

This one works:

```
curl http://api.archivenode.io/<I_ALMOST_POSTED_MY_API_KEY>  -H "Content-Type: application/json" -d '{"id":1,"jsonrpc":"2.0","method":"eth_call","params":[{"data":"0x9b013aee000000000000000000000000000000000000000000000000000000000000000a","to":"0x542156d51D10Db5acCB99f9Db7e7C91B74E80a2c"},10428450]}'
```

Returns:

```
{"jsonrpc":"2.0","id":1,"result":"0x000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"}
```

If you run the same call at just one block later, it will hang.

```
curl http://api.archivenode.io/<I_ALMOST_POSTED_MY_API_KEY>  -H "Content-Type: application/json" -d '{"id":1,"jsonrpc":"2.0","method":"eth_call","params":[{"data":"0x9b013aee000000000000000000000000000000000000000000000000000000000000000a","to":"0x542156d51D10Db5acCB99f9Db7e7C91B74E80a2c"},10428451]}'
```

The reason for that is that this is calling into a tokenset contract and the first call is made when the contract is in rebalancing state whereas the second one is not and then the call will revert. But instead of printing out the revert reason, the API call just hangs.

**Comment by tkstanczak** [comment]
For now just pasting the observed behaviour.
We will check now the Geth behaviour on it.

supernode@neth-supernode:~$ curl mainnet-supernode.nethermind.io  -H "Content-Type: application/json" -d '{"id":1,"jsonrpc":"2.0","method":"eth_call","params":[{"data":"0x9b013aee000000000000000000000000000000000000000000000000000000000000000a","to":"0x542156d51D10Db5acCB99f9Db7e7C91B74E80a2c"},10428450]}'
{"jsonrpc":"2.0","result":"0x000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000","id":1}

supernode@neth-supernode:~$ curl https://mainnet-supernode.nethermind.io  -H "Content-Type: application/json" -d '{"id":1,"jsonrpc":"2.0","method":"eth_call","params":[{"data":"0x9b013aee000000000000000000000000000000000000000000000000000000000000000a","to":"0x542156d51D10Db5acCB99f9Db7e7C91B74E80a2c"},10428451]}'
{"jsonrpc":"2.0","error":{"code":-32015,"message":"VM execution error.","data":"revert: Invalid state"},"id":1}

**Comment by cburgdorf** [comment]
Interesting! It seems that the call at block `10428451` is working for you and you get the revert reason as expected. That's not the case for me when I try it with my api.achivenode URL.

**Comment by tkstanczak** [comment]
ok, I think we will need to analyze the logs of the api.archivenode

**Comment by MysticRyuujin** [comment]
What version is mainnet-supernode?

I tried the call against our nodes directly (without `dshackle`, without any front-end) and they all timeout.

(you guys can't hit these endpoints, it's firewall by IP address, but if you ping me in discord I can allow your IP to hit it directly)

**Comment by tkstanczak** [comment]
[discord] <MysticRyuujin> If I curl it from localhost...
curl http://localhost:8545 -H "Content-Type: application/json" -d '{"id":1,"jsonrpc":"2.0","method":"eth_call","params":[{"data":"0x9b013aee000000000000000000000000000
> 000000000000000000000000000000000000a","to":"0x542156d51D10Db5acCB99f9Db7e7C91B74E80a2c"},10428451]}'

{"jsonrpc":"2.0","error":{"code":-32015,"message":"VM execution error.","data":"revert"},"id":1}

**Comment by tkstanczak** [comment]
![image](https://user-images.githubusercontent.com/498913/87553791-3f2f6780-c6ab-11ea-9e66-fa9602c4faeb.png)

**Comment by tkstanczak** [comment]
![image](https://user-images.githubusercontent.com/498913/87553938-6ede6f80-c6ab-11ea-91af-0dc239d7553f.png)

**Comment by MysticRyuujin** [comment]
Alright so an update:

My timeouts directly to the node were due to an IPS rule...because my calls directly to the node were over HTTP and in plain text, the Intrusion Prevention System was freaking out.

The calls over HTTPS are not seen by the IPS but they do go through `dshackle`...and `dshackle` is also throwing an error. An error that I *THOUGHT* was due to the error messages I saw in the Nethermind node, but that seems to be a "normal" error response to this situation...and it's `dshackle` that is not handling that error response correctly.

I have opened an issue with `dshackle` over `eth_call revert` errors not being handled correctly...we'll see what he says.

With all of that said, i think this issue can be closed, and it would make more sense to open a NEW issue specifically for MyCrypto integration...because even bypassing `dshackle` still results in MyCrypto being unable to return token balances, and this has nothing to do with Archive node data 😄
